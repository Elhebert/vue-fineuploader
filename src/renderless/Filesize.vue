<script>
const DEFAULT_UNITS = ['byte', 'kilobyte', 'megabyte', 'gigabyte', 'terabyte']

export default {
  props: {
    id: {
      type: Number,
      required: true,
    },
    uploader: {
      type: Object,
      required: true,
    },
  },

  data: () => ({
    size: 0,
    onUpload: null,
  }),

  computed: {
    formatted() {
      let formatSizeAndUnits = {}
      let size = this.size

      Object.values(DEFAULT_UNITS).some(unit => {
        if (size >= 1e3) {
          size = (size / 1e3).toFixed(2)

          return false
        }

        formatSizeAndUnits = { size, unit }
        return true
      })

      return formatSizeAndUnits
    },
  },

  created() {
    this.size = this.uploader.methods.getSize(this.id)

    const scalingOption = this.uploader.options.scaling

    if (scalingOption && scalingOption.sizes.length) {
      this.onUpload = id => {
        if (!this.isCurrentFile(id)) {
          return
        }

        this.size = this.uploader.methods.getSize(id)
      }
    }
  },

  mounted() {
    this.onUpload && this.uploader.on('upload', this.onUpload)
  },

  beforeDestroy() {
    this.onUpload && this.uploader.off('upload', this.onUpload)
  },

  methods: {
    isCurrentFile(id) {
      return this.id === id
    },
  },

  render() {
    return this.$scopedSlots.default({
      size: this.formatted.size,
      unit: this.formatted.unit,
    })
  },
}
</script>
