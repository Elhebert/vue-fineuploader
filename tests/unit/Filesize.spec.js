import { mount } from '@vue/test-utils'
import Filesize from '@/Filesize.vue'
import FineUploader from 'fine-uploader-wrappers'

describe('Filesize.vue', () => {
  const nativeObjectToString = Object.prototype.toString

  beforeEach(() => {
    Object.prototype.toString = function() {
      if (this && this.type === 'fakeBlob') {
        return '[object Blob]'
      }

      return nativeObjectToString.apply(this, arguments)
    }
  })

  afterEach(() => {
    Object.prototype.toString = nativeObjectToString
  })

  it('renders an empty filesize component if size is not known initially', () => {
    const uploader = new FineUploader({ options: { autoUpload: false } })
    uploader.methods.addFiles({ type: 'fakeBlob' })

    const wrapper = mount(Filesize, {
      propsData: { id: 0, uploader },
    })

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.text()).toBe('')
  })

  it('renders formatted file size for various sized files', () => {
    const uploader = new FineUploader({ options: { autoUpload: false } })

    uploader.methods.addFiles([
      { size: 1100, type: 'fakeBlob' },
      { size: 1100000, type: 'fakeBlob' },
      { size: 1100000000, type: 'fakeBlob' },
      { size: 1100000000000, type: 'fakeBlob' },
    ])

    const expectedSizes = [
      { size: '1.10', units: 'KB' },
      { size: '1.10', units: 'MB' },
      { size: '1.10', units: 'GB' },
      { size: '1.10', units: 'TB' },
    ]

    expectedSizes.forEach((expectedSize, id) => {
      const wrapper = mount(Filesize, {
        propsData: { id, uploader },
      })

      expect(wrapper).toMatchSnapshot()
      expect(wrapper.text()).toBe(`${expectedSize.size} ${expectedSize.units}`)
    })
  })

  it('renders file size at upload time for scaled blobs', () => {
    const uploader = new FineUploader({
      options: {
        autoUpload: false,
        scaling: { sizes: [{ name: 'test', maxSize: 100 }] },
      },
    })
    uploader.methods.addFiles({ type: 'fakeBlob' })

    const wrapper = mount(Filesize, {
      propsData: { id: 0, uploader },
    })

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.text()).toBe('')

    uploader.methods.getSize = jest.fn(() => 1)

    const renderless = wrapper.vm.$children[0]
    renderless.onUpload(0)

    expect(uploader.methods.getSize).toHaveBeenCalledWith(0)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.text()).toEqual('1 B')
  })
})
