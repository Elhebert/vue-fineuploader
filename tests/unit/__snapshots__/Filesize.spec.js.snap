// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filesize.vue renders an empty filesize component if size is not known initially 1`] = ``;

exports[`Filesize.vue renders file size at upload time for scaled blobs 1`] = ``;

exports[`Filesize.vue renders file size at upload time for scaled blobs 2`] = `<span>1 B</span>`;

exports[`Filesize.vue renders formatted file size for various sized files 1`] = `<span>1.10 KB</span>`;

exports[`Filesize.vue renders formatted file size for various sized files 2`] = `<span>1.10 MB</span>`;

exports[`Filesize.vue renders formatted file size for various sized files 3`] = `<span>1.10 GB</span>`;

exports[`Filesize.vue renders formatted file size for various sized files 4`] = `<span>1.10 TB</span>`;
